flink/flink-1.14/src/main/java/io/tidb/bigdata/flink/connector/sink/TiDBDataStreamSinkProvider.java

```diff
@@ -0,0 +1,274 @@
+import static io.tidb.bigdata.flink.connector.TiDBOptions.SinkTransaction.MINIBATCH;
+
+import io.tidb.bigdata.flink.connector.TiDBOptions.SinkTransaction;
+import io.tidb.bigdata.flink.connector.sink.function.TiDBKeyedProcessFunctionFactory;
+import io.tidb.bigdata.flink.connector.sink.function.TiDBSinkFunction;
+import io.tidb.bigdata.flink.connector.sink.operator.TiDBCommitOperator;
+import io.tidb.bigdata.flink.connector.sink.operator.TiDBGlobalWriteOperator;
+import io.tidb.bigdata.flink.connector.sink.operator.TiDBMiniBatchWriteOperator;
+import io.tidb.bigdata.flink.connector.sink.operator.TiDBWriteOperator;
+import io.tidb.bigdata.flink.connector.sink.serializer.TiDBTransactionContextSerializer;
+import io.tidb.bigdata.flink.connector.sink.serializer.TiDBTransactionStateSerializer;
+import io.tidb.bigdata.flink.connector.utils.TiDBRowConverter;
+import io.tidb.bigdata.tidb.ClientConfig;
+import io.tidb.bigdata.tidb.ClientSession;
+import io.tidb.bigdata.tidb.SqlUtils;
+import io.tidb.bigdata.tidb.TiDBWriteHelper;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.DataStreamSink;
+import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
+import org.apache.flink.streaming.api.environment.CheckpointConfig;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.catalog.ResolvedCatalogTable;
+import org.apache.flink.table.connector.sink.DataStreamSinkProvider;
+import org.apache.flink.table.connector.sink.DynamicTableSink.Context;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.utils.TableSchemaUtils;
+import org.apache.flink.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.tikv.common.BytePairWrapper;
+import org.tikv.common.meta.TiColumnInfo;
+import org.tikv.common.meta.TiIndexColumn;
+import org.tikv.common.meta.TiIndexInfo;
+import org.tikv.common.meta.TiTableInfo;
+import org.tikv.common.meta.TiTimestamp;
+import org.tikv.common.row.Row;
+
+public class TiDBDataStreamSinkProvider implements DataStreamSinkProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TiDBDataStreamSinkProvider.class);
+  public static final String PRIMARY_KEY_PREFIX = "TIBIGDATA_PRIMARY_KEY_";
+
+  private final ResolvedCatalogTable table;
+  private final Context context;
+  private final String databaseName;
+  private final String tableName;
+  private final Map<String, String> properties;
+  private final TableSchema tableSchema;
```

Comment:
`TableSchema` has been deprecated, do we need to change it to `Schema`

Reply:
> `TableSchema` has been deprecated, do we need to change it to `Schema`

Yes, it would be better use `Schema`, but I found some interesting things that many of the code of Flink still uses `TableSchema`. It seems that this class uses too much and will not be abandoned for a while. We can do it in next PR.

Reply:
Yes, I have tried to replace `TableSchema` with `schema`, but found it's not a simple replace.

Reply:
> 
agree